### ISSUE TITLE
Change datatype of date fields in MSSQL

### ISSUE THREAD

**Issue by nollymar** [dotCMS/core#21169]
On [this card](https://github.com/dotCMS/core/issues/20899) , the way dates were persisted was modified to consider time zones in Postgres. To finish this work, we need to do the same in MSSQL

**Comment by fabrizzio-dotCMS** [comment]
Apparently, this is what we need to change date types to 
https://docs.microsoft.com/en-us/sql/t-sql/data-types/datetimeoffset-transact-sql?view=sql-server-ver15

**Comment by nollymar** [comment]
**Internal QA:** Needs Work

Running an upgrade in MSSQL from 22.03 to 22.06 fails multiple times with this exception: 
```
dotcms-dev_1     | 23:31:06.086  ERROR runonce.Task220331UpdateDateTimezones - error executing: ALTER TABLE dist_reindex_journal ALTER COLUMN time_entered datetimeoffset(3)
dotcms-dev_1     | com.microsoft.sqlserver.jdbc.SQLServerException: The object 'DF__dist_rein__time___43A1090D' is dependent on column 'time_entered'.
dotcms-dev_1     | 	at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:262) ~[mssql-jdbc-7.4.1.jre8.jar:?]
dotcms-dev_1     | 	at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1624) ~[mssql-jdbc-7.4.1.jre8.jar:?]
dotcms-dev_1     | 	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatement(SQLServerPreparedStatement.java:594) ~[mssql-jdbc-7.4.1.jre8.jar:?]
dotcms-dev_1     | 	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtExecCmd.doExecute(SQLServerPreparedStatement.java:524) ~[mssql-jdbc-7.4.1.jre8.jar:?]
dotcms-dev_1     | 	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7194) ~[mssql-jdbc-7.4.1.jre8.jar:?]
dotcms-dev_1     | 	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:2979) ~[mssql-jdbc-7.4.1.jre8.jar:?]
dotcms-dev_1     | 	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:248) ~[mssql-jdbc-7.4.1.jre8.jar:?]
dotcms-dev_1     | 	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:223) ~[mssql-jdbc-7.4.1.jre8.jar:?]
dotcms-dev_1     | 	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.execute(SQLServerPreparedStatement.java:505) ~[mssql-jdbc-7.4.1.jre8.jar:?]
dotcms-dev_1     | 	at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44) ~[HikariCP-3.4.2.jar:?]
dotcms-dev_1     | 	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java) ~[HikariCP-3.4.2.jar:?]
dotcms-dev_1     | 	at com.dotmarketing.startup.runonce.Task220331UpdateDateTimezones.lambda$updateTable$c1f54200$1(Task220331UpdateDateTimezones.java:47) ~[dotcms_22.06_999999.jar:?]
dotcms-dev_1     | 	at io.vavr.control.Try.of(Try.java:75) ~[vavr-0.10.3.jar:?]
dotcms-dev_1     | 	at com.dotmarketing.startup.runonce.Task220331UpdateDateTimezones.updateTable(Task220331UpdateDateTimezones.java:47) ~[dotcms_22.06_999999.jar:?]
dotcms-dev_1     | 	at com.dotmarketing.startup.runonce.Task210901UpdateDateTimezones.executeUpgrade(Task210901UpdateDateTimezones.java:82) ~[dotcms_22.06_999999.jar:?]
dotcms-dev_1     | 	at com.dotmarketing.startup.StartupTasksExecutor.executeUpgrades(StartupTasksExecutor.java:199) ~[dotcms_22.06_999999.jar:?]
dotcms-dev_1     | 	at com.liferay.portal.servlet.MainServlet.init(MainServlet.java:119) ~[dotcms_22.06_999999.jar:?]
dotcms-dev_1     | 	at org.apache.catalina.core.StandardWrapper.initServlet(StandardWrapper.java:1164) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.StandardWrapper.loadServlet(StandardWrapper.java:1117) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.StandardWrapper.load(StandardWrapper.java:1010) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.StandardContext.loadOnStartup(StandardContext.java:4957) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5264) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:726) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:698) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:696) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.startup.HostConfig.deployDirectory(HostConfig.java:1185) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:1933) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
dotcms-dev_1     | 	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
dotcms-dev_1     | 	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75) ~[tomcat-util.jar:9.0.60]
dotcms-dev_1     | 	at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:118) ~[?:?]
dotcms-dev_1     | 	at org.apache.catalina.startup.HostConfig.deployDirectories(HostConfig.java:1095) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:477) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.startup.HostConfig.start(HostConfig.java:1618) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:319) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.util.LifecycleBase.fireLifecycleEvent(LifecycleBase.java:123) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.util.LifecycleBase.setStateInternal(LifecycleBase.java:423) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.util.LifecycleBase.setState(LifecycleBase.java:366) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:946) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:835) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1396) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1386) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
dotcms-dev_1     | 	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75) ~[tomcat-util.jar:9.0.60]
dotcms-dev_1     | 	at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:140) ~[?:?]
dotcms-dev_1     | 	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:919) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.StandardEngine.startInternal(StandardEngine.java:263) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.StandardService.startInternal(StandardService.java:432) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.core.StandardServer.startInternal(StandardServer.java:927) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.startup.Catalina.start(Catalina.java:772) ~[catalina.jar:9.0.60]
dotcms-dev_1     | 	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
dotcms-dev_1     | 	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
dotcms-dev_1     | 	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
dotcms-dev_1     | 	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
dotcms-dev_1     | 	at org.apache.catalina.startup.Bootstrap.start(Bootstrap.java:345) ~[bootstrap.jar:9.0.60]
dotcms-dev_1     | 	at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:476) ~[bootstrap.jar:9.0.60]
```

Besides, the max(db_version) remains with the task consecutive  `220331` after the failure. It should have rolled back to `220330`.

**Comment by fabrizzio-dotCMS** [comment]
As far as I recall in spite of all the reported errors the upgrade task succeeds. 
That's why the version number shows as the last.
I stumbled into the same situation ( yeah I forgot to document it) but then again as far as I recall all the target columns are migrated correctly to the new expected target type.

**Comment by nollymar** [comment]
**Internal QA:** Passed

After running an upgrade from 22.03 to 22.06, everything works as expected and date/time fields are of type `datetimeoffset`

**Comment by bryanboza** [comment]
Fixed, tested on release-22.06 // Docker // FF

**Comment by erickgonzalez** [comment]
Only for 22.03 LTS
